src/adapter/src/catalog/open.rs

```diff
@@ -465,77 +477,7 @@ impl Catalog {
                 }
             }
 
-            let clusters = txn.get_clusters();
-            let mut cluster_azs = BTreeMap::new();
-            let builtin_log_lookup: BTreeMap<_, _> = BUILTINS::logs().map(|log| (log.name, log)).collect();
-            for mz_catalog::durable::Cluster {
-                id,
-                name,
-                owner_id,
-                privileges,
-                config,
-            } in clusters
-            {
-                let introspection_source_index_ids = txn.get_introspection_source_indexes(id);
-                let all_indexes: Vec<_> = introspection_source_index_ids.into_iter()
-                    // TODO(jkosh44) There may be some old deleted logs stored durably that no longer
-                    // exists. For now we ignore them, but we should clean them up.
-                    .filter_map(|(log_name, (id, oid))| {
-                        let log = builtin_log_lookup.get(log_name.as_str());
-                        log.map(|log| (*log, id, oid))
-                    })
-                    .collect();
-
-                if let mz_catalog::durable::ClusterVariant::Managed(managed) = &config.variant {
-                    cluster_azs.insert(id, managed.availability_zones.clone());
-                }
-
-                state.insert_cluster(
-                    id,
-                    name,
-                    all_indexes,
-                    owner_id,
-                    PrivilegeMap::from_mz_acl_items(privileges),
-                    config.into(),
-                );
-            }
-
-            let replicas = txn.get_cluster_replicas();
-            let mut allocated_replicas = Vec::new();
-            for mz_catalog::durable::ClusterReplica {
-                cluster_id,
-                replica_id,
-                name,
-                config,
-                owner_id,
-            } in replicas
-            {
-                let logging = ReplicaLogging {
-                    log_logging: config.logging.log_logging,
-                    interval: config.logging.interval,
-                };
-                let config = ReplicaConfig {
-                    location: state.concretize_replica_location(
-                        config.location,
-                        &vec![],
-                        cluster_azs.get(&cluster_id).map(|zones| &**zones),
-                    )?,
-                    compute: ComputeReplicaConfig {
-                        logging,
-                    },
-                };
-
-                allocated_replicas.push(mz_catalog::durable::ClusterReplica {
-                    cluster_id,
-                    replica_id,
-                    name: name.clone(),
-                    config: config.clone().into(),
-                    owner_id: owner_id.clone(),
-                });
-
-                state.insert_cluster_replica(cluster_id, name, replica_id, config, owner_id);
-            }
-            txn.set_replicas(allocated_replicas)?;
```

Comment:
I think this is something @guswynn wrote. Iirc, we have to account for a situation where the availability zones change after an envd restart, but I might be mistaken.

Reply:
Let me know your thoughts @guswynn. Looking at the code, it looks like `cluster_azs` is only populated from the persisted config of all clusters and doesn't look at some global list of all availability zones.

Reply:
@jkosh44 I think this 
1. asserting that any replica attached to a cluster does not have an az configuration 
2. ensuring `MANAGED` replica configs's are generated with a correct `availability` (and maybe `disk`?) based on the managed _cluster_ configuration. This is required because the k8s orchestrator is only told about the replicas, not the top-level cluster

Reply:
I think handling changes to the `--availability-zone` configuration is entirely handled by `Catalog::open` and `ensure_valid_azs`

Reply:
> asserting that any replica attached to a cluster does not have an az configuration

Are you sure about this? Maybe I'm misunderstanding, but from `concretize_replica_location`, it looks like it's valid for the replica to have an az configuration, regardless of what the cluster az configuration is.

https://github.com/MaterializeInc/materialize/blob/9267be261a245de04e8fce8eb1a2054f98b3a039/src/adapter/src/catalog/state.rs#L2057

> ensuring `MANAGED` replica configs's are generated with a correct `availability` (and maybe `disk`?) based on the managed cluster configuration. This is required because the k8s orchestrator is only told about the replicas, not the top-level cluster

...

> I think handling changes to the `--availability-zone` configuration is entirely handled by `Catalog::open` and `ensure_valid_azs`

`concretize_replica_location` does not change anything about the `disk` field as far as I can tell. If I'm understanding correctly, you're saying that this code is responsible for setting the replica's `availability_zone` field to the correct value in the follow scenarios:
  - When the replica is first created.
  - When we reboot and the `--availability-zone` configuration has changed.

Does that sound right? If so,

> When the replica is first created.

This looks right to me, but implies that we only need to run this once, when the replica is first created, not every time we reboot.

> When we reboot and the `--availability-zone` configuration has changed.

I am not able to see where this happens. It doesn't look like `Catalog::open` ever calls `ensure_valid_azs` and also it never reads the value of `--availability-zone` other than to set a field in `CatalogState`.

Reply:
Chatted with @guswynn offline and we decided that `txn.set_replicas(allocated_replicas)?;` is not needed since `concretize_replica_location` should never change the durable representation of the replica.